File: frontend/src/api/versioning.ts
import client from "./client";
import type { Branch, Version, VersionDiff } from "../types";

export interface ListVersionsParams {
  branchId?: string;
  limit?: number;
  offset?: number;
  search?: string;
  createdAfter?: string;
  createdBefore?: string;
}

export interface VersionUpdateRequest {
  is_pinned?: boolean;
}

export interface SnapshotCreateRequest {
  name?: string;
  description?: string;
  branch_id?: string;
}

export interface RestoreRequest {
  handle_extra_images: "keep" | "remove";
  pre_restore_snapshot: boolean;
}

export interface RatingImpact {
  /** Live images whose rating the restore would change, in either direction. */
  will_change: number;
  /** The subset it would *clear* — rated after the snapshot was taken. */
  will_clear: number;
  /** Rated images the version does not contain. Under `handle_extra_images:
   *  "remove"` these are **deleted**, not reverted, so the modal reports them
   *  separately: which number applies depends on the mode selected. */
  extras_rated: number;
}

export interface RestoreSummary {
  files_restored: number;
  files_unavailable: number;
  files_failed: number;
  images_re_created: number;
  images_removed: number;
  pre_restore_version_id: string | null;
}

export const versioningApi = {
  listBranches: (datasetId: string) =>
    client.get<Branch[]>(`/datasets/${datasetId}/versions/branches`).then((r) => r.data),

  createBranch: (datasetId: string, name: string, fromVersionId?: string, includeSnapshot = true) =>
    client
      .post<Branch | { job_id: string }>(`/datasets/${datasetId}/versions/branches`, {
        name,
        from_version_id: fromVersionId,
        include_snapshot: includeSnapshot,
      })
      .then((r) => r.data),

  checkoutBranch: (datasetId: string, branchId: string, preRestoreSnapshot = true) =>
    client
      .post<{ job_id: string }>(`/datasets/${datasetId}/versions/branches/${branchId}/checkout`, {
        pre_restore_snapshot: preRestoreSnapshot,
      })
      .then((r) => r.data),

  listVersions: (datasetId: string, params: ListVersionsParams = {}) =>
    client
      .get<Version[]>(`/datasets/${datasetId}/versions`, {
        params: {
          branch_id: params.branchId,
          limit: params.limit ?? 50,
          offset: params.offset ?? 0,
          search: params.search || undefined,
          created_after: params.createdAfter || undefined,
          created_before: params.createdBefore || undefined,
        },
      })
      .then((r) => r.data),

  createSnapshot: (datasetId: string, body: SnapshotCreateRequest) =>
    client
      .post<Version | { job_id: string }>(`/datasets/${datasetId}/versions`, body)
      .then((r) => r.data),

  getVersion: (datasetId: string, versionId: string) =>
    client.get<Version>(`/datasets/${datasetId}/versions/${versionId}`).then((r) => r.data),

  deleteBranch: (datasetId: string, branchId: string) =>
    client.delete(`/datasets/${datasetId}/versions/branches/${branchId}`),

  deleteVersion: (datasetId: string, versionId: string) =>
    client.delete(`/datasets/${datasetId}/versions/${versionId}`),

  updateVersion: (datasetId: string, versionId: string, body: VersionUpdateRequest) =>
    client
      .patch<Version>(`/datasets/${datasetId}/versions/${versionId}`, body)
      .then((r) => r.data),

  /** What restoring this version would do to the keep/cut ratings on disk now.
   *  A version-vs-*current* comparison, which `diffVersions` is not — that reads
   *  two snapshots. A rating is hand-made work nothing recomputes, so the
   *  restore confirm states the number first. */
  ratingImpact: (datasetId: string, versionId: string) =>
    client
      .get<RatingImpact>(`/datasets/${datasetId}/versions/${versionId}/rating-impact`)
      .then((r) => r.data),

  restoreVersion: (datasetId: string, versionId: string, body: RestoreRequest) =>
    client
      .post<{ job_id: string }>(`/datasets/${datasetId}/versions/${versionId}/restore`, body)
      .then((r) => r.data),

  pruneStorage: (datasetId: string) =>
    client
      .post<{ job_id: string }>(`/datasets/${datasetId}/versions/prune`)
      .then((r) => r.data),

  diff: (datasetId: string, v1: string, v2: string) =>
    client
      .get<VersionDiff>(`/datasets/${datasetId}/versions/diff`, { params: { v1, v2 } })
      .then((r) => r.data),
};

File: backend/schemas/versioning.py
from typing import Any, Literal

from pydantic import BaseModel

from backend.schemas import UtcDatetime


class BranchOut(BaseModel):
    id: str
    dataset_id: str
    name: str
    head_version_id: str | None
    head_version_name: str | None = None
    created_at: UtcDatetime

    model_config = {"from_attributes": True}


class BranchCreate(BaseModel):
    name: str
    from_version_id: str | None = None
    include_snapshot: bool = True


class CheckoutRequest(BaseModel):
    pre_restore_snapshot: bool = True


class VersionOut(BaseModel):
    id: str
    dataset_id: str
    branch_id: str | None
    parent_id: str | None
    name: str | None
    description: str
    image_count: int
    created_at: UtcDatetime
    source: Literal["manual", "pre_restore", "branch_init"]
    is_pinned: bool

    model_config = {"from_attributes": True}


class SnapshotCreate(BaseModel):
    name: str | None = None
    description: str = ""
    branch_id: str | None = None


class VersionUpdate(BaseModel):
    is_pinned: bool | None = None


class RestoreRequest(BaseModel):
    handle_extra_images: str = "keep"
    pre_restore_snapshot: bool = True


class RestoreSummary(BaseModel):
    files_restored: int
    files_unavailable: int
    files_failed: int = 0
    images_re_created: int
    images_removed: int
    pre_restore_version_id: str | None


class ImageDiffEntry(BaseModel):
    image_id: str | None
    filename: str
    subfolder: str
    caption: str


class ModifiedImageDiff(BaseModel):
    image_id: str | None
    filename: str
    subfolder: str
    changes: dict[str, Any]


class DiffOut(BaseModel):
    added: list[ImageDiffEntry]
    removed: list[ImageDiffEntry]
    modified: list[ModifiedImageDiff]
    unchanged_count: int
    summary: dict[str, int]


class RatingImpact(BaseModel):
    """What restoring a version would do to the keep/cut ratings on disk now.

    Genuinely a version-vs-*current* comparison, which nothing else here does —
    `diff_versions` reads `VersionImageState` on both sides. A restore reverts a
    rating like any other mirrored column, and a rating is hand-made work that no
    job can recompute, so the confirm dialog says how much of it is at stake.
    """
    # Live images whose rating the restore would change (in either direction).
    will_change: int
    # The subset it would *clear* — a rating given after the snapshot was taken.
    will_clear: int
    # Rated images the version does not contain. Under handle_extra_images="remove"
    # these rows are **deleted**, not reverted, so the modal counts them apart:
    # which of the two numbers applies depends on the mode the user has picked.
    extras_rated: int
